url47ebb27197_asirbu_CorrectiveEvolution_tgz-pJ8-eu_fbk_soa_evolution_engine_impl_test_ProblemToSMVFuncTestJ8/src/eu/fbk/soa/eventlog/LogExplorer.java

```diff
@@ -0,0 +1,202 @@
+package eu.fbk.soa.eventlog;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.log4j.Logger;
+import org.deckfour.xes.extension.std.XConceptExtension;
+import org.deckfour.xes.extension.std.XLifecycleExtension;
+import org.deckfour.xes.info.XLogInfo;
+import org.deckfour.xes.info.impl.XLogInfoImpl;
+import org.deckfour.xes.model.XEvent;
+import org.deckfour.xes.model.XLog;
+import org.deckfour.xes.model.XTrace;
+
+import eu.fbk.soa.process.Activity;
+import eu.fbk.soa.process.ProcessModel;
+import eu.fbk.soa.process.Trace;
+
+
+public class LogExplorer {
+
+	static Logger logger = Logger.getLogger(LogExplorer.class);
+	
+	private int distinctTraceNr = 0;
+	
+	private int totalNrTraces;
+	
+	private ProcessModel model;
+	
+	private XConceptExtension conceptExt;
+	
+	private XLifecycleExtension lifecycleExt;
+	
+	private Map<String, List<String>> traceTypes;
+	
+	private Set<String> divergingTraces;
+	
+	private Set<String> modelTraces;
+	
+	private Map<String, Integer> nrOfTracesByType;
+	
+	private Map<String, Boolean> canBeReplayed;
+	
+	
+	public LogExplorer(ProcessModel model) {
+		this.model = model;	
+		this.modelTraces = new HashSet<String>();
+		conceptExt = XConceptExtension.instance();
+		lifecycleExt = XLifecycleExtension.instance();
+	}
+	
+	private void init() {
+		canBeReplayed = new HashMap<String, Boolean>();
+		traceTypes = new HashMap<String, List<String>>();
+		nrOfTracesByType = new HashMap<String, Integer>();
+		divergingTraces = new HashSet<String>();
+	}
+	
+	
+	public void processEventLog(XLog log) {
+		init();
+		processTracesInOriginalModel();
+		
+		XLogInfo info = XLogInfoImpl.create(log);
+		totalNrTraces = info.getNumberOfTraces();
+		
+		logger.info("Processing event log: found " + 
+				info.getNumberOfTraces() + " traces and a total of " + info.getNumberOfEvents() + " events");
+		
+		for (XTrace trace : log) {
+			processTrace(trace);			
+		}
+		logger.info("Identified " + traceTypes.size() + " distinct traces, " +
+				"of which " + divergingTraces.size() + " cannot be replayed by model");
+	}
+	
+	private void processTracesInOriginalModel() {
+		for (Trace trace : model.getDistinctTraces()) {
+			distinctTraceNr++;
+			String newTraceID = "" + distinctTraceNr;
+			
+			List<String> actNames = new ArrayList<String>();
+			for (Activity act : trace.getActivities()) {
+				actNames.add(act.getName());
+			}
+			traceTypes.put(newTraceID, actNames);
+			this.nrOfTracesByType.put(newTraceID, 0);
+			this.modelTraces.add(newTraceID);
+			
+			logger.debug("Added original trace: " + actNames);
+		}
+	}
+	
+	private void processTrace(XTrace trace) {
+//		System.out.println("Processing trace " + index);
+		List<String> traceActivities = this.getActivities(trace);
+		
+		String traceID = this.getTraceType(traceActivities);
+		if (traceID != "") {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.